HotelBase/HotelBaseClass.cs

```diff
@@ -54,5 +58,54 @@ internal bool BookRoomInExternalService(DateTime date)
         {
             return _bookingService.Book(date);
         }
+
+        public ReservationResult Reserve(DateTime date, double price, int creditCardNumber, string email)
+        {
+            bool paymentMade = false;
+            bool roomBooked = false;
+            bool emailSent = false;
+            bool success = true;
+            bool priceValid = false;
+            string reservationNumber = null;
+
+            while (Operations.Count != 0)
+            {
+                HotelOperation currentOperation = Operations.OrderBy(o => o.Order).First();
+                bool stepSuccess = true;
+
+                switch (currentOperation.Operation)
+                {
+                    case Operation.BookRoom:
+                        roomBooked = BookRoom(date);
+                        stepSuccess = roomBooked;
+                        break;
+                    case Operation.CheckPrice:
+                        priceValid = CheckPrice(price, date);
+                        stepSuccess = priceValid;
+                        break;
+                    case Operation.MakePayment:
+                        paymentMade = MakePayment(creditCardNumber, price);
+                        stepSuccess = paymentMade;
+                        break;
+                    case Operation.SendEmail:
+                        emailSent = SendEmail(email);
+                        stepSuccess = emailSent;
+                        break;
+                    case Operation.GenerateReservationNumber:
+                        reservationNumber = GenerateReservationNumber();
+                        break;
+                }
+
+                if (!stepSuccess && currentOperation.ShouldFailWholeProcess)
+                {
+                    success = false;
+                    break;
+                }
+
+                Operations.Remove(currentOperation);
```

Comment:
Same like above with foreach it will not be a need for this

Reply:
Applied